### PR TITLE
bug fix to allow random interval untranslated hails

### PR DESCRIPTION
**Pull request by UnorderedSigh**
**Bugfix:** Fixes a bug (issue #8037) that prevented untranslated random interval hails.

## Fix Details

1. Untranslated random interval hails work again.
2. Systems with lots of ships that *don't* have untranslated hails and a few that do, will get those hails more often. This is because, when randomly deciding a ship to send hails, the game only considers ones that *can send hails*. This was a big problem for the Kor Efret system since most random interval selected the Quarg, so you rarely got Kor Efret hails.
3. This will *not* cause stations and ship hail screens to let you talk to untranslatable ships. (I tested that.) The untranslated bug fix *only* applies to random interval hails. If you look in my code changes, you will see that CanSendHail has an optional argument, allowUntranslated, which defaults to false. It is only set to true for random interval hails.

## Testing Done
Tested talking to an untranslated station or planet, to confirm it doesn't respond.

## Performance Impact
Inconsequential.

### CODE REVIEW

**Comment by Ferociousfeind** [comment]
Relatively related... can we at some point in the future get "prompted" v. "unprompted" hails? Trying to write text that fits both situations is awkward and difficult. You can't have pirates say "what you lookin' at?" because I'm not looking at a damn thing, mister! And hails being generally unprompted-style like "My passengers won't shut up about this new band, There Might Be Riots." is awkward when you specifically hail them and they unprompted start ranting about TMBR.

**Comment by UnorderedSigh** [comment]
> Relatively related... can we at some point in the future get "prompted" v. "unprompted" hails? Trying to write text that fits both situations is awkward and difficult. You can't have pirates say "what you lookin' at?" because I'm not looking at a damn thing, mister! And hails being generally unprompted-style like "My passengers won't shut up about this new band, There Might Be Riots." is awkward when you specifically hail them and they unprompted start ranting about TMBR.

You get such hails by selecting a ship and pressing "t" but I haven't fixed those yet. I want to get random interval hails working before I try to separate hailing from services in the hail panel.

**Comment by UnorderedSigh** [comment]
>  This simply means ships can send untranslated hails to you in space but will not change actual hails when you hail a ship/planet.

There is an easy bugfix to allow hailing an untranslated ship without allowing access to services. I haven't implemented it, since I don't want to put any more time into untranslated hails until I know if they'll be allowed at all. Planets are a bit more work, since there is no way to specify untranslated planet text right now.

File: source/Engine.cpp [hunk 1]
```diff
@@ -155,8 +155,10 @@ namespace {
 				|| ship->Cloaking() >= 1. || ship->GetPersonality().IsMute())
 			return false;
 
-		// Ships that don't share a language with the player shouldn't send hails.
-		if(!gov->Language().empty() && !player.Conditions().Get("language: " + gov->Language()))
+		// Ships that don't share a language with the player shouldn't communicate normally.
```
Comment (warp-core) [inline source/Engine.cpp:158]:
```suggestion
		// Ships that don't share a language with the player shouldn't communicate when hailed directly.
```

File: source/Engine.cpp [hunk 2]
```diff
@@ -1782,18 +1784,22 @@ void Engine::SendHails()
...
+	vector<shared_ptr<const Ship>> canSend;
+	canSend.reserve(ships.size());
+
+	// When deciding who will send a hail, only consider ships that can send hails.
+	for(auto &it : ships)
+		if(CanSendHail(it, player, true))
+			canSend.push_back(it);
 
-	if(!CanSendHail(source, player))
+	if(!canSend.size())
```
Comment (Hurleveur) [inline source/Engine.cpp:1795]:
```suggestion
	if(!canSend.empty())
```

**Review by warp-core — APPROVED** [review]

**Pull request merged**